### PR TITLE
GridGeoSampler: don't change stride of last patch

### PR DESCRIPTION
**Pull request by adamjstewart**
Partially reverts #630. See #1245 for motivation.

@remtav

### CODE REVIEW

**Comment by isaaccorley** [comment]
Are we sure this works? I thought that to sample outside the bounds of an image we have to pass `boundless=True` like

```python
dest = src.read(
   indexes=band_indexes,
   out_shape=out_shape,
   window=from_bounds(*bounds, src.transform),
   boundless=True
)
```

**Comment by adamjstewart** [comment]
Good catch! It returns an image but crops it to the bounds of the file unless boundless=True is added. I'll fix this.

File: torchgeo/datasets/geo.py [hunk 1]
```diff
@@ -487,6 +487,7 @@ def _merge_files(
                 indexes=band_indexes,
                 out_shape=out_shape,
                 window=from_bounds(*bounds, src.transform),
+                boundless=True,
             )
         else:
             dest, _ = rasterio.merge.merge(
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:493]: `merge` has no `boundless` parameter, but doesn't need one. I confirmed that it properly samples outside the bounds of the file and uses nodata pixels for padding.

**Review by calebrob6 — APPROVED** [review]

**Comment by adamjstewart** [comment]
This introduces a serious bug. IntersectionDatasets with different CRSs result in the following error when loading data:
```
Traceback (most recent call last):
  File "/Users/Adam/torchgeo/test.py", line 11, in <module>
    ds[ds.bounds]
  File "/Users/Adam/torchgeo/torchgeo/datasets/geo.py", line 899, in __getitem__
    samples = [ds[query] for ds in self.datasets]
  File "/Users/Adam/torchgeo/torchgeo/datasets/geo.py", line 899, in <listcomp>
    samples = [ds[query] for ds in self.datasets]
  File "/Users/Adam/torchgeo/torchgeo/datasets/geo.py", line 443, in __getitem__
    data = self._merge_files(filepaths, query, self.band_indexes)
  File "/Users/Adam/torchgeo/torchgeo/datasets/geo.py", line 486, in _merge_files
    dest = src.read(
  File "rasterio/_warp.pyx", line 1233, in rasterio._warp.WarpedVRTReaderBase.read
ValueError: WarpedVRT does not permit boundless reads
```
The fix is to use `rasterio.merge.merge` for all I/O instead.

**Pull request merged**